### PR TITLE
Add python-standalone target

### PR DESCRIPTION
**Pull request by clinssen**
Replaces https://github.com/nest/nestml/pull/729.

See the added documentation (.rst) for a full description.

A ``ModelPrinter`` class was introduced as a superclass for ``ASTPrinter``s that are expected to be able to print complete models. NEST-specific AST utils were split into a different utils file.

### CODE REVIEW

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/ReadFromDictionaryToTmp.jinja2 [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{#
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/ReadFromDictionaryToTmp.jinja2:1]: Is this file required?

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/ReturnStatement.jinja2 [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{#
+  Generates a single return statement in C++ syntax.
+  @param: ast A single ast-return stmt object. ASTReturnStmt
+#}
+{%- if tracing %}/* generated by {{self._TemplateReference__context.name}} */ {% endif %}
+{%- if ast.has_expression() %}
+return {{printer.print_expression(ast.get_expression())}};
+{%- else %}
+return;
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/ReturnStatement.jinja2:9]:
```suggestion
return {{printer.print_expression(ast.get_expression())}}
{%- else %}
return
```

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/SmallStatement.jinja2 [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{#
+  Generates a single small statement into equivalent C++ syntax.
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/SmallStatement.jinja2:2]:
```suggestion
  Generates a single small statement into equivalent Python syntax.
```

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/StateVariablesEnum.jinja2 [hunk 4]
```diff
@@ -0,0 +1,20 @@
+{#
+  Generates an Enum with state variables that are recordable when the neuron model uses vectors
+  @param neuron ASTNeuron: the neuron model
+-#}
+
+enum StateVecVars {
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/StateVariablesEnum.jinja2:6]: Are vectors supported for the python-standalone target?

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/Statement.jinja2 [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{#
+  Generates a single statement, either a simple or compound, to equivalent C++ syntax.
+  @param ast ASTSmallStmt or ASTCompoundStmt
+#}
+{%- if tracing %}/* generated by {{self._TemplateReference__context.name}} */ {% endif %}
+{%- if stmt.has_comment() %}
+{{stmt.print_comment('//')}}{%- endif %}
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/Statement.jinja2:7]:
```suggestion
{{stmt.print_comment('#')}}{%- endif %}
```

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/WhileStatement.jinja2 [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{#
+  Generates C++ declaration
+  @param ast ASTWhileStmt
+#}
+{%- if tracing %}/* generated by {{self._TemplateReference__context.name}} */ {% endif %}
+while ( {{printer.print_expression(ast.get_condition())}})
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/WhileStatement.jinja2:6]: This needs to be changed with the python syntax.

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/WriteInDictionary.jinja2 [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{#
+  Generates an instruction to write a value into the dictionary of registered variables.
+  @param variable VariableSymbol
+#}
+{%- if tracing %}/* generated by {{self._TemplateReference__context.name}} */ {% endif %}
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/WriteInDictionary.jinja2:5]: Is this file required?

File: doc/running.rst [hunk 8]
```diff
@@ -163,3 +186,11 @@ To generate code that is compatible with particular release versions of NEST Sim
 - The default is the empty string, which causes the NEST version to be automatically identified from the ``nest`` Python module.
 - ``"v2.20.2"``: Latest NEST 2 release.
 - ``"master"``: Latest NEST GitHub master branch version (https://github.com/nest/nest-simulator/).
+
+
+Python-standalone target
+------------------------
+
+*NESTML features supported: neurons*
```
Comment (pnbabu) [inline doc/running.rst:194]: Could you also mention the features supported currently for the python-standalone target? It would also be beneficial to mention the list of files generated for the target and how to run the simulation after.

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/StateVariablesEnum.jinja2 (hunk 4 above)
Comment (clinssen) [reply pynestml/codegeneration/resources_python_standalone/point_neuron/directives/StateVariablesEnum.jinja2:6]: Probably this would not be very difficult to implement as indeed a lot of the prerequisites are already there, but it would require adding at least some unit tests, if it's alright with you I'd like to save that for a separate PR.

**Comment by clinssen** [comment]
Thanks for the extensive review! The new commits should address all of your comments.

You can see the rendered documentation at: https://nestml-sandbox.readthedocs.io/en/latest/running.html

File: doc/running.rst [hunk 9]
```diff
@@ -163,3 +186,30 @@ To generate code that is compatible with particular release versions of NEST Sim
...
+
+
+Python-standalone target
+------------------------
+
+*NESTML features supported:* :doc:`neurons </nestml_language/neurons_in_nestml>`
+
+The aim of the Python-standalone target is to facilitate model development and debugging. A Python class is generated for each neuron, as well as a very simple simulator that applies some spikes to the model(s) and measures the results. This generated code can be run independently of any installed simulator (only a few common Python packages are required, like scipy for numerical integration). The generated Python code is intended to be easy to read and understand, rather than to be fast. When satisfied with the Python target results, high-performance code can then be generated by simply switching to a different target platform.
+
+.. list-table::
```
Comment (pnbabu) [inline doc/running.rst:198]: Could you also add how to run this? I suspect it's just `python test_python_standalone_module.py`?

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/GSLDifferentiationFunction.jinja2 [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+{%- endfor %}
+{%- if paired_synapse is defined %}
+{%- for variable_name in numeric_state_variables_moved %}
+{%-   set update_expr = numeric_update_expressions[variable_name] %}
+{%-   set variable_symbol = neuron.get_scope().resolve_to_symbol(variable_name, SymbolKind.VARIABLE) %}
+  f[node.S_.ode_state_variable_name_to_index["{{variable_symbol.name}}"]] = {{gsl_printer.print(update_expr)}}
+{%- endfor %}
+{%- endif %}
+
+  return f
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/GSLDifferentiationFunction.jinja2:36]:
```suggestion
  return f
  
```

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/MemberDeclaration.jinja2 [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+#}
+{%- if tracing %}# generated by {{self._TemplateReference__context.name}}{% endif %}
+{%- if variable_symbol.has_comment() %}
+{{ variable_symbol.print_comment("# ") }}
+{%- endif %}
+{%- if variable_symbol.get_symbol_name() in numeric_state_variables %}
+ode_state[ode_state_variable_name_to_index["{{ printer_no_origin.print(variable) }}"]] = np.nan  # type: {{ declarations.print_variable_type(variable_symbol) }}
+{%- else %}
+{{ printer_no_origin.print(variable) }} = None  # type: {{ declarations.print_variable_type(variable_symbol) }}
+{%- endif %}
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/MemberDeclaration.jinja2:14]:
```suggestion
{%- endif %}

```

**Review by pnbabu — COMMENTED** [review]
Some minor comments. Looks good to me otherwise.

**Comment by pnbabu** [comment]
This doesn't have the updated doc: https://nestml-sandbox.readthedocs.io/en/latest/running.html

File: doc/running.rst (hunk 9 above)
Comment (clinssen) [reply doc/running.rst:206]: I added a shell command example just after this paragraph, could you have a look? Cheers!

**Comment by pnbabu** [comment]
> > This doesn't have the updated doc: https://nestml-sandbox.readthedocs.io/en/latest/running.html
> 
> Please refresh!

It works now, but the `nest` and `python` logos in [this page](https://nestml-sandbox.readthedocs.io/en/latest/running.html) are too big. Is it possible to make them smaller so that they are placed just beside their respective headings?

File: pynestml/codegeneration/resources_python_standalone/point_neuron/directives/GSLDifferentiationFunction.jinja2 [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
```
Comment (pnbabu) [inline pynestml/codegeneration/resources_python_standalone/point_neuron/directives/GSLDifferentiationFunction.jinja2:37]: I see a new line here but I don't know why it still says no new line.

File: doc/running.rst (hunk 9 above)
Comment (pnbabu) [reply doc/running.rst:206]: This looks good to me. Thanks!

**Comment by clinssen** [comment]
> > > This doesn't have the updated doc: https://nestml-sandbox.readthedocs.io/en/latest/running.html
> > 
> > 
> > Please refresh!
> 
> It works now, but the `nest` and `python` logos in [this page](https://nestml-sandbox.readthedocs.io/en/latest/running.html) are too big. Is it possible to make them smaller so that they are placed just beside their respective headings?

This should be fixed now!

**Review by pnbabu — APPROVED** [review]
Thanks!
